### ISSUE TITLE
scatterplot error with 3 pts, a NaN, and an RGB color

### ISSUE THREAD

**Issue by tritemio** [matplotlib/matplotlib#8658]
### Bug report

**Bug summary**

When importing `seaborn` a simple scatterplot containing NaNs fails with

```
ValueError: Invalid RGBA argument: 0.29
```
Here is the code for reproducing the issue:

```python
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import seaborn
matplotlib.__version__

%matplotlib inline

plt.scatter([1, 2, 3], [1, 2, np.nan])
```

**Actual outcome**

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/colors.py in to_rgba(c, alpha)
    140     try:
--> 141         rgba = _colors_full_map.cache[c, alpha]
    142     except (KeyError, TypeError):  # Not in cache, or unhashable.

KeyError: (0.29803921568627451, None)

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/colors.py in _to_rgba_no_colorcycle(c, alpha)
    191     try:
--> 192         c = tuple(map(float, c))
    193     except TypeError:

TypeError: 'numpy.float64' object is not iterable

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-7-e1ec5a517bd0> in <module>()
      7 get_ipython().magic('matplotlib inline')
      8 
----> 9 plt.scatter([1, 2, 3], [1, 2, np.nan])

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/pyplot.py in scatter(x, y, s, c, marker, cmap, norm, vmin, vmax, alpha, linewidths, verts, edgecolors, hold, data, **kwargs)
   3433                          vmin=vmin, vmax=vmax, alpha=alpha,
   3434                          linewidths=linewidths, verts=verts,
-> 3435                          edgecolors=edgecolors, data=data, **kwargs)
   3436     finally:
   3437         ax._hold = washold

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/__init__.py in inner(ax, *args, **kwargs)
   1890                     warnings.warn(msg % (label_namer, func.__name__),
   1891                                   RuntimeWarning, stacklevel=2)
-> 1892             return func(ax, *args, **kwargs)
   1893         pre_doc = inner.__doc__
   1894         if pre_doc is None:

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/axes/_axes.py in scatter(self, x, y, s, c, marker, cmap, norm, vmin, vmax, alpha, linewidths, verts, edgecolors, **kwargs)
   4026                 offsets=offsets,
   4027                 transOffset=kwargs.pop('transform', self.transData),
-> 4028                 alpha=alpha
   4029                 )
   4030         collection.set_transform(mtransforms.IdentityTransform())

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/collections.py in __init__(self, paths, sizes, **kwargs)
    888         """
    889 
--> 890         Collection.__init__(self, **kwargs)
    891         self.set_paths(paths)
    892         self.set_sizes(sizes)

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/collections.py in __init__(self, edgecolors, facecolors, linewidths, linestyles, antialiaseds, offsets, transOffset, norm, cmap, pickradius, hatch, urls, offset_position, zorder, **kwargs)
    137         self._is_filled = True  # May be modified by set_facecolor().
    138 
--> 139         self.set_facecolor(facecolors)
    140         self.set_edgecolor(edgecolors)
    141         self.set_linewidth(linewidths)

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/collections.py in set_facecolor(self, c)
    672         """
    673         self._original_facecolor = c
--> 674         self._set_facecolor(c)
    675 
    676     def set_facecolors(self, c):

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/collections.py in _set_facecolor(self, c)
    657         except AttributeError:
    658             pass
--> 659         self._facecolors = mcolors.to_rgba_array(c, self._alpha)
    660         self.stale = True
    661 

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/colors.py in to_rgba_array(c, alpha)
    235     result = np.empty((len(c), 4), float)
    236     for i, cc in enumerate(c):
--> 237         result[i] = to_rgba(cc, alpha)
    238     return result
    239 

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/colors.py in to_rgba(c, alpha)
    141         rgba = _colors_full_map.cache[c, alpha]
    142     except (KeyError, TypeError):  # Not in cache, or unhashable.
--> 143         rgba = _to_rgba_no_colorcycle(c, alpha)
    144         try:
    145             _colors_full_map.cache[c, alpha] = rgba

/Users/anto/miniconda3/envs/py36/lib/python3.6/site-packages/matplotlib/colors.py in _to_rgba_no_colorcycle(c, alpha)
    192         c = tuple(map(float, c))
    193     except TypeError:
--> 194         raise ValueError("Invalid RGBA argument: {!r}".format(orig_c))
    195     if len(c) not in [3, 4]:
    196         raise ValueError("RGBA sequence should have length 3 or 4")

ValueError: Invalid RGBA argument: 0.29803921568627451
```
**Expected outcome**

Plot a scatterplot with 3 points.

**Matplotlib version**
  * Operating System: mac os / conda
  * Matplotlib Version: 2.0.0 and 2.0.2
  * Python Version: 3.6
  * Jupyter Version (if applicable): 5
  * Other Libraries: seaborn

**Comment by efiring** [comment]
I see where this is coming from, and it is actually not Seaborn's fault, so I will change the title.  It is a problem with specifying a color as an RGB triplet, which Seaborn does in setting the default prop cycle.  In the context of scatter's horrendously complex handling of all the possible ways that colors can be specified, with or without mapping, this fails in the example with three points including one nan, but not 3 points without the nan:
```python
import matplotlib.pyplot as plt
plt.scatter([1, 2, 3], [4, 2, 2], color=(0.5, 0.3, 0))  # OK
plt.scatter([1, 2, 3], [4, 2, np.nan], color=(0.5, 0.3, 0))  # fails
```
This is not a simple matter of the inevitable ambiguity between mapping 3 colors and specifying a single RGB; if it were, the first line would yield 3 colors, which it does not.

I think I see how to make a simple change that would fix the original problem with the Seaborn import, and maybe the example above, but a better solution would be something I wanted to do long ago, but never got around to doing: make genuine classes for a color and an array of colors (or maybe just the latter, with a single color being one case), so that once something is determined to be a color or an array of colors, it retains that identity ever after, or at least within the Python parts of the code.

**Comment by tritemio** [comment]
@efiring, thanks! Just a note: the problem is there also with 2 NaNs.

**Comment by efiring** [comment]
Correct.  Note to myself: the problem is in the following block, given the argument manipulations leading up to it:
```python
        # Anything in maskargs will be unchanged unless it is the same length
        # as x:
        maskargs = x, y, s, c, colors, edgecolors, linewidths
        x, y, s, c, colors, edgecolors, linewidths =\
            cbook.delete_masked_points(*maskargs)
```
Colors (and edgecolors, if used), when given as an RGB, is being trimmed as if it was an array of colors.

**Comment by phobson** [comment]
Just for refererence: https://github.com/mwaskom/seaborn/issues/840

**Comment by has2k1** [comment]
This bug is not present in version 2.0.2.post4242+gb6eb043de. It may have been fixed by accident.

**Comment by tacaswell** [comment]
Closing as this is reported fixed on 2.1.x
